crates/common/src/lib.rs

```diff
@@ -534,6 +535,48 @@ pub fn calculate_class_commitment_leaf_hash(
     )
 }
 
+pub fn deployed_contract_address(
+    constructor_calldata: impl Iterator<Item = CallParam>,
+    contract_address_salt: &ContractAddressSalt,
+    class_hash: &ClassHash,
+) -> ContractAddress {
+    let constructor_calldata_hash = constructor_calldata
+        .fold(HashChain::default(), |mut h, param| {
+            h.update(param.0);
+            h
+        })
+        .finalize();
+
+    let contract_address = [
+        Felt::from_be_slice(b"STARKNET_CONTRACT_ADDRESS").expect("prefix is convertible"),
+        Felt::ZERO,
+        contract_address_salt.0,
+        class_hash.0,
+        constructor_calldata_hash,
+    ]
+    .into_iter()
+    .fold(HashChain::default(), |mut h, e| {
+        h.update(e);
+        h
+    })
+    .finalize();
+
+    // Contract addresses are _less than_ 2**251 - 256
+    let contract_address = primitive_types::U256::from_big_endian(contract_address.as_be_bytes());
+    let max_address = primitive_types::U256::from_str_radix(
+        "0x7ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00",
+        16,
+    )
+    .unwrap();
+
+    let contract_address = contract_address.rem(max_address);
+    let mut b = [0u8; 32];
+    contract_address.to_big_endian(&mut b);
+    let contract_address = Felt::from_be_slice(&b).unwrap();
```

Comment:
How often will we be doing these calcs? This feels expensive - or is `U256` big-endian based?

Reply:
No, `U256` is little-endian... Ideally we could do this calculation with Felts so that there's no need for using an U256 just for the modulo division. Maybe there's a way to do that.

Reply:
I'm afraid it's not. Wrt p2p parsing: all of these computations are now done in bulk in a rayon thread pool.

Reply:
We can implement `std::ops::Sub` for Felt and convert the remainder operation to a subtraction:

```rust
impl std::ops::Sub for Felt {
    type Output = Felt;

    fn sub(self, rhs: Self) -> Self::Output {
        let result = MontFelt::from(self) - MontFelt::from(rhs);
        Felt::from(result)
    }
}
```

And:

```rust
// Contract addresses are _less than_ 2**251 - 256
const MAX_CONTRACT_ADDRESS: Felt =
    felt!("0x7ffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff00");
let contract_address = if contract_address >= MAX_CONTRACT_ADDRESS {
    contract_address - MAX_CONTRACT_ADDRESS
} else {
    contract_address
};
```

Reply:
Will do it now :arrow_up: